File: packages/core/src/data/downsample.ts
/**
 * Largest-Triangle-Three-Buckets (LTTB) downsampling — reduce a series to
 * `threshold` points while preserving its visual shape (peaks/troughs), the
 * standard technique for plotting very long line series. First and last points
 * are always kept.
 */
export function lttb(
  x: ArrayLike<number>, y: ArrayLike<number>, threshold: number,
): { x: Float64Array; y: Float64Array } {
  const n = Math.min(x.length, y.length);
  if (threshold >= n || threshold <= 2) {
    const rx = new Float64Array(n), ry = new Float64Array(n);
    for (let i = 0; i < n; i++) { rx[i] = x[i]!; ry[i] = y[i]!; }
    return { x: rx, y: ry };
  }

  const sx = new Float64Array(threshold), sy = new Float64Array(threshold);
  const bucket = (n - 2) / (threshold - 2);
  let a = 0; // index of the previously-kept point
  sx[0] = x[0]!; sy[0] = y[0]!;
  let out = 1;

  for (let i = 0; i < threshold - 2; i++) {
    // Average of the next bucket (the triangle's third vertex).
    let avgX = 0, avgY = 0;
    let rs = Math.floor((i + 1) * bucket) + 1;
    const re = Math.min(Math.floor((i + 2) * bucket) + 1, n);
    const rlen = re - rs || 1;
    for (let k = rs; k < re; k++) { avgX += x[k]!; avgY += y[k]!; }
    avgX /= rlen; avgY /= rlen;

    // Point in this bucket forming the largest triangle with `a` and the average.
    let cs = Math.floor(i * bucket) + 1;
    const ce = Math.floor((i + 1) * bucket) + 1;
    const ax = x[a]!, ay = y[a]!;
    let maxArea = -1, chosen = cs, cx = x[cs]!, cy = y[cs]!;
    for (; cs < ce; cs++) {
      const area = Math.abs((ax - avgX) * (y[cs]! - ay) - (ax - x[cs]!) * (avgY - ay));
      if (area > maxArea) { maxArea = area; chosen = cs; cx = x[cs]!; cy = y[cs]!; }
    }
    sx[out] = cx; sy[out] = cy; out++;
    a = chosen;
  }

  sx[out] = x[n - 1]!; sy[out] = y[n - 1]!;
  return { x: sx, y: sy };
}

File: native/src/data/downsample.cpp
#include "data/downsample.hpp"

#include <algorithm>
#include <cmath>

namespace photon::data {

Downsampled lttb(const double* x, const double* y, size_t count, size_t threshold) {
  Downsampled out;
  if (!x || !y) count = 0;
  if (threshold >= count || threshold <= 2) {
    out.x.assign(x, x + count);
    out.y.assign(y, y + count);
    return out;
  }

  out.x.assign(threshold, 0.0);
  out.y.assign(threshold, 0.0);
  const double bucket = static_cast<double>(count - 2) / static_cast<double>(threshold - 2);
  size_t previous = 0;  // index of the last point kept
  out.x[0] = x[0];
  out.y[0] = y[0];
  size_t at = 1;

  for (size_t i = 0; i + 2 < threshold; ++i) {
    // The next bucket's average is the triangle's third vertex.
    const size_t range_start =
        static_cast<size_t>(std::floor(static_cast<double>(i + 1) * bucket)) + 1;
    const size_t range_end = std::min(
        static_cast<size_t>(std::floor(static_cast<double>(i + 2) * bucket)) + 1, count);
    const double span = static_cast<double>(range_end > range_start ? range_end - range_start : 1);
    double avg_x = 0.0;
    double avg_y = 0.0;
    for (size_t k = range_start; k < range_end; ++k) {
      avg_x += x[k];
      avg_y += y[k];
    }
    avg_x /= span;
    avg_y /= span;

    // The point in this bucket that makes the largest triangle with the last
    // kept point and that average — which is what preserves a spike.
    size_t cursor = static_cast<size_t>(std::floor(static_cast<double>(i) * bucket)) + 1;
    const size_t stop = static_cast<size_t>(std::floor(static_cast<double>(i + 1) * bucket)) + 1;
    const double ax = x[previous];
    const double ay = y[previous];
    double max_area = -1.0;
    size_t chosen = cursor;
    for (; cursor < stop && cursor < count; ++cursor) {
      const double area =
          std::abs((ax - avg_x) * (y[cursor] - ay) - (ax - x[cursor]) * (avg_y - ay));
      if (area > max_area) {
        max_area = area;
        chosen = cursor;
      }
    }
    out.x[at] = x[chosen];
    out.y[at] = y[chosen];
    ++at;
    previous = chosen;
  }

  out.x[at] = x[count - 1];
  out.y[at] = y[count - 1];
  return out;
}

}  // namespace photon::data
